File: examples/python/obstacle_data_recorder/obstacle_data_recorder/obstacle_data_recorder.py
import os
import sys

import zmq

try:
    from zmq_msgs.obstacle_data import ObstacleData
    from zmq_msgs.topic_ports import OBSTACLE_TOPIC
except ImportError:
    sys.path.append(
        os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../../zmq_msgs/python"))
    )
    from zmq_msgs.obstacle_data import ObstacleData
    from zmq_msgs.topic_ports import OBSTACLE_TOPIC


def write_data(filename, obstacle_data):
    with open(filename, "w") as out:
        out.write("x1,z1,x2,z2,x3,z3,x4,z4,vx,vz\n")
        for obstacle in obstacle_data.obstacles:
            for p in obstacle.bounding_box.points:
                out.write(f"{p.x:.6f},{p.z:.6f},")
            out.write(f"{obstacle.velocity.x:.6f},{obstacle.velocity.z:.6f}\n")


class ObstacleDataSink:
    def __init__(self, endpoint, output_dir):
        self.context = zmq.Context(1)
        self.socket = self.context.socket(zmq.SUB)
        self.socket.setsockopt(zmq.RCVHWM, 1)
        self.socket.setsockopt_string(zmq.SUBSCRIBE, "")
        self.socket.connect(endpoint)
        self.last_frame_id = 0
        print(f"Subscribing to {endpoint}")
        self.output_dir = output_dir
        os.makedirs(output_dir, exist_ok=True)

    def loop_once(self):
        msg = self.socket.recv()
        obstacle_data = ObstacleData()
        obstacle_data.read(msg)
        frame_id = obstacle_data.frame_id

        # Warn if we dropped a frame
        if self.last_frame_id != 0 and frame_id != self.last_frame_id + 1:
            print(
                f"{frame_id - self.last_frame_id - 1} frames dropped. "
                f"Current frame ID: {frame_id}, last frame ID: {self.last_frame_id}"
            )
        self.last_frame_id = frame_id

        filename = os.path.join(self.output_dir, f"{frame_id:09}.txt")
        print(f"\rWriting {filename}", end="", flush=True)
        write_data(filename, obstacle_data)


def print_usage(default_ip, default_output_dir):
    print(
        "You should specify the IP address of the device running Hammerhead,\n"
        "as well as the folder where you want the data to be saved:\n\n"
        "     python obstacle_data_recorder.py hammerhead_ip output_dir\n\n"
        "e.g. python obstacle_data_recorder.py 192.168.1.9 obstacle_datas\n\n"
        "If unspecified, then we assume that you are running this on the device running Hammerhead,\n"
        "along with the other defaults\n\n"
        f"     python obstacle_data_recorder.py {default_ip} {default_output_dir}\n"
        "----------------------------------------"
    )


def main():
    default_ip = "127.0.0.1"
    topic = OBSTACLE_TOPIC
    default_output_dir = "obstacle_datas"
    if len(sys.argv) < 3:
        print_usage(default_ip, default_output_dir)
    ip = sys.argv[1] if len(sys.argv) > 1 else default_ip

    output_dir = sys.argv[2] if len(sys.argv) >= 3 else default_output_dir
    endpoint = f"tcp://{ip}:{topic.port}"
    sink = ObstacleDataSink(endpoint, output_dir)
    while True:
        sink.loop_once()


if __name__ == "__main__":
    main()

File: examples/cpp/obstacle_data_recorder/src/obstacle_data_recorder.cpp
#include <atomic>
#include <csignal>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <nodar/zmq/obstacle_data.hpp>
#include <nodar/zmq/topic_ports.hpp>
#include <zmq.hpp>

std::atomic_bool running{true};

void signalHandler(int signum) {
    std::cerr << "SIGINT or SIGTERM received." << std::endl;
    running = false;
}

inline void write_data(const std::string &filename, const nodar::zmq::ObstacleData &obstacleData) {
    std::ofstream out{filename};
    out << "x1," << "z1," << "x2," << "z2," << "x3," << "z3," << "x4," << "z4," << "vx," << "vz" << "\n";
    for (const auto &obstacle : obstacleData.obstacles) {
        for (const auto &p : obstacle.boundingBox.points) {
            out << std::fixed << std::setprecision(6) << p.x << "," << p.z << ",";
        }
        out << obstacle.velocity.x << "," << obstacle.velocity.z << '\n';
    }
    out << std::endl;
}

class ObstacleDataSink {
public:
    uint64_t last_frame_id = 0;

    ObstacleDataSink(const std::filesystem::path &output_dir, const std::string &endpoint)
        : output_dir(output_dir), context(1), socket(context, ZMQ_SUB) {
        const int hwm = 1;  // set maximum queue length to 1 message
        socket.set(zmq::sockopt::rcvhwm, hwm);
        socket.set(zmq::sockopt::subscribe, "");
        socket.connect(endpoint);
        std::cout << "Subscribing to " << endpoint << std::endl;
    }

    void loopOnce() {
        zmq::message_t msg;
        const auto received_bytes = socket.recv(msg, zmq::recv_flags::none);
        const nodar::zmq::ObstacleData obstacleData{static_cast<uint8_t *>(msg.data())};
        const auto &frame_id = obstacleData.frame_id;

        // Warn if we dropped a frame
        if ((last_frame_id != 0) and (frame_id != last_frame_id + 1)) {
            std::cerr << (frame_id - last_frame_id - 1) << " frames dropped. Current frame ID : " << frame_id
                      << ", last frame ID: " << last_frame_id << std::endl;
        }
        last_frame_id = frame_id;
        std::cout << "\rFrame # " << frame_id << ". " << std::flush;

        std::ostringstream filename_ss;
        filename_ss << std::setw(9) << std::setfill('0') << frame_id << ".txt";
        const auto filename = output_dir / filename_ss.str();
        std::cout << "Writing " << filename << std::flush;
        write_data(filename, obstacleData);
    }

private:
    std::filesystem::path output_dir;
    zmq::context_t context;
    zmq::socket_t socket;
};

void printUsage(const std::string &default_ip) {
    std::cout << "You should specify the IP address of the device running hammerhead:\n\n"
                 "     ./obstacle_data_recorder hammerhead_ip\n\n"
                 "e.g. ./obstacle_data_recorder 10.10.1.10\n\n"
                 "In the meantime, we assume that you are running this on the device running Hammerhead,\n"
                 "that is, we assume that you specified\n\n"
                 "     ./obstacle_data_recorder "
              << default_ip << "\n----------------------------------------" << std::endl;
}

int main(int argc, char *argv[]) {
    static constexpr auto default_ip = "127.0.0.1";
    static constexpr auto topic = nodar::zmq::OBSTACLE_TOPIC;
    signal(SIGINT, signalHandler);
    signal(SIGTERM, signalHandler);
    if (argc == 1) {
        printUsage(default_ip);
    }
    const auto ip{argc > 1 ? argv[1] : default_ip};
    const auto endpoint{std::string("tcp://") + ip + ":" + std::to_string(topic.port)};

    const auto HERE = std::filesystem::path(__FILE__).parent_path();
    const auto output_dir = HERE / "obstacle_datas";
    std::filesystem::create_directories(output_dir);

    ObstacleDataSink sink(output_dir, endpoint);
    while (running) {
        sink.loopOnce();
    }
}